### ISSUE TITLE
improper handling of include+with_item resets variable

### ISSUE THREAD

**Issue by h0nIg** [ansible/ansible#17180]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

include
##### ANSIBLE VERSION

2.1.1.0
##### CONFIGURATION

N/A
##### OS / ENVIRONMENT

N/A
##### SUMMARY

the second include of testinclude1.yml should output overriddenval and not testval. This happens only if the with_item is used on the include of testinclude2.yml. I think there is some variable magic happening regarding the block statement, because the problem will not occur if testinclude3.yml is used instead of testinclude1.yml. This problem will **NOT** occur with 2.1.0.0
##### STEPS TO REPRODUCE

```

---
- hosts: XXX

  vars:
    testvar1: testval

  tasks:
    - include: testinclude1.yml

    - include: testinclude2.yml
      with_items:
        - testitem

    - include: testinclude1.yml
```

**testinclude1.yml**

```
- block:
  - debug: msg="{{ testvar1 }}"
  vars:
    testvar1: overriddenval
```

**testinclude2.yml**

```
- debug: msg="test"
```

**testinclude3.yml**

```
- debug: msg="{{ testvar1 }}"
  vars:
    testvar1: overriddenval
```
##### EXPECTED RESULTS

```
TASK [setup] *******************************************************************
ok: [XXX]

TASK [debug] *******************************************************************
ok: [XXX] => {
    "msg": "overriddenval"
}

TASK [include] *****************************************************************
included: XXX/testinclude2.yml for XXX

TASK [debug] *******************************************************************
ok: [XXX] => {
    "msg": "test"
}

TASK [debug] *******************************************************************
ok: [XXX] => {
    "msg": "overriddenval"
}
```
##### ACTUAL RESULTS

```
TASK [setup] *******************************************************************
ok: [XXX]

TASK [debug] *******************************************************************
ok: [XXX] => {
    "msg": "overriddenval"
}

TASK [include] *****************************************************************
included: XXX/testinclude2.yml for XXX

TASK [debug] *******************************************************************
ok: [XXX] => {
    "msg": "test"
}

TASK [debug] *******************************************************************
ok: [XXX] => {
    "msg": "testval"
}
```

**Comment by alikins** [comment]
I reproduce the failure on 2.1.1.0 and git stable-2.1 branch. 
It seems to work on 2.1.0.0 and devel (2.2*):

PASSING on 2.1.0

```
[fedora-23:include_override_17180 (v2.1.0.0-1-branch %)]$ ansible --version
ansible 2.1.0.0 (detached HEAD ref: refs/) last updated 2016/08/22 12:25:17 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 91a839f1e3) last updated 2016/08/22 12:23:14 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 1aeb9f8a8c) last updated 2016/08/22 12:23:14 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = ['/home/adrian/ansible/my-modules', '/home/adrian/ansible/lib/modules/']
[fedora-23:include_override_17180 (v2.1.0.0-1-branch %)]$ ansible-playbook -vv test.yml 
Using /home/adrian/.ansible.cfg as config file

PLAYBOOK: test.yml *************************************************************
1 plays in test.yml

PLAY [swoop] *******************************************************************

TASK [setup] *******************************************************************
ok: [swoop]

TASK [include] *****************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/test.yml:8
included: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml for swoop

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml:2
ok: [swoop] => {
    "msg": "overriddenval"
}

TASK [include] *****************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/test.yml:10
included: /home/adrian/src/ansible/repro/include_override_17180/testinclude2.yml for swoop

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude2.yml:1
ok: [swoop] => {
    "msg": "test"
}

TASK [include] *****************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/test.yml:14
included: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml for swoop

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml:2
ok: [swoop] => {
    "msg": "overriddenval"
}

PLAY RECAP *********************************************************************
swoop                      : ok=7    changed=0    unreachable=0    failed=0   

```

FAILING on v2.1.1.0:

```
[fedora-23:include_override_17180 (v2.1.1.0-1-branch %)]$ ansible --version
ansible 2.1.1.0 (v2.1.1.0-1-branch 780c363482) last updated 2016/08/19 10:12:11 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 91a839f1e3) last updated 2016/08/22 12:23:14 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 1aeb9f8a8c) last updated 2016/08/22 12:23:14 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = ['/home/adrian/ansible/my-modules', '/home/adrian/ansible/lib/modules/']
[fedora-23:include_override_17180 (v2.1.1.0-1-branch %)]$ ansible-playbook -vv test.yml 
Using /home/adrian/.ansible.cfg as config file

PLAYBOOK: test.yml *************************************************************
1 plays in test.yml

PLAY [swoop] *******************************************************************

TASK [setup] *******************************************************************
ok: [swoop]

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml:2
ok: [swoop] => {
    "msg": "overriddenval"
}

TASK [include] *****************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/test.yml:10
included: /home/adrian/src/ansible/repro/include_override_17180/testinclude2.yml for swoop

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude2.yml:1
ok: [swoop] => {
    "msg": "test"
}

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml:2
ok: [swoop] => {
    "msg": "testval"
}

PLAY RECAP *********************************************************************
swoop                      : ok=5    changed=0    unreachable=0    failed=0   

```

FAILING on stable-2.1

```
[fedora-23:include_override_17180 (stable-2.1 % u=)]$ ansible --version
ansible 2.1.1.0 (stable-2.1 bf929ac532) last updated 2016/08/22 12:19:47 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 493ead2d42) last updated 2016/08/22 12:19:52 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 5a84de640b) last updated 2016/08/22 12:19:52 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = ['/home/adrian/ansible/my-modules', '/home/adrian/ansible/lib/modules/']
[fedora-23:include_override_17180 (stable-2.1 % u=)]$ ansible-playbook -vv test.yml 
Using /home/adrian/.ansible.cfg as config file

PLAYBOOK: test.yml *************************************************************
1 plays in test.yml

PLAY [swoop] *******************************************************************

TASK [setup] *******************************************************************
ok: [swoop]

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml:2
ok: [swoop] => {
    "msg": "overriddenval"
}

TASK [include] *****************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/test.yml:10
included: /home/adrian/src/ansible/repro/include_override_17180/testinclude2.yml for swoop

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude2.yml:1
ok: [swoop] => {
    "msg": "test"
}

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml:2
ok: [swoop] => {
    "msg": "testval"
}

PLAY RECAP *********************************************************************
swoop                      : ok=5    changed=0    unreachable=0    failed=0   


```

But I get the expected results from the report if I test against devel/:
ie, PASSING on devel

```
[fedora-23:include_override_17180 (devel % u=)]$ ansible --version
ansible 2.2.0 (devel a4785c2691) last updated 2016/08/22 12:23:14 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 91a839f1e3) last updated 2016/08/22 12:23:14 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 1aeb9f8a8c) last updated 2016/08/22 12:23:14 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = ['/home/adrian/ansible/my-modules', '/home/adrian/ansible/lib/modules/']
[fedora-23:include_override_17180 (devel % u=)]$ ansible-playbook -vv test.yml 
Using /home/adrian/.ansible.cfg as config file
statically included: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml
statically included: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml

PLAYBOOK: test.yml *************************************************************
1 plays in test.yml

PLAY [swoop] *******************************************************************

TASK [setup] *******************************************************************
ok: [swoop]

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml:2
ok: [swoop] => {
    "msg": "overriddenval"
}

TASK [include] *****************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/test.yml:10
included: /home/adrian/src/ansible/repro/include_override_17180/testinclude2.yml for swoop

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude2.yml:1
ok: [swoop] => {
    "msg": "test"
}

TASK [debug] *******************************************************************
task path: /home/adrian/src/ansible/repro/include_override_17180/testinclude1.yml:2
ok: [swoop] => {
    "msg": "overriddenval"
}

PLAY RECAP *********************************************************************
swoop                      : ok=5    changed=0    unreachable=0    failed=0   
```

**Comment by h0nIg** [comment]
@jimi-c @alikins did you find the reason for this?

**Comment by h0nIg** [comment]
same behavior on the following:

**playbook:**

```
---
- hosts: testhost
  vars:
    testvar1: overriddenval

  roles:
    - {role: testrole, testvar1: testval}
```

**testrole:**

```
- debug: msg="{{ testvar1 }}"
  vars:
    testvar1: overriddenval
```

**output:**

```
TASK [testrole : debug] ************************************************************
ok: [test] => {
    "msg": "testval"
}
```

**expected output:**

```
TASK [testrole : debug] ************************************************************
ok: [test] => {
    "msg": "overriddenval"
}
```

**Comment by h0nIg** [comment]
maybe related to the following issues:
https://github.com/ansible/ansible/issues/14411
https://github.com/ansible/ansible/pull/14826
https://github.com/ansible/ansible/issues/14702
https://github.com/ansible/ansible/issues/16729

**Comment by jimi-c** [comment]
There was a bug introduced in 2.1.1 I believe, relating to static vs. dynamic includes, which was fixed in devel. Unfortunately that fix was related to the performance improvements work we were doing recently, and it's not back-portable to the stable-2.1 branch. I'll continue looking into this to see if there's a different way to fix it.

**Comment by h0nIg** [comment]
@jimi-c look at the second example, there is no include that is used so this is not related to the problem you described. could you include this into 2.1.2?

**Comment by h0nIg** [comment]
@jimi-c @alikins still failing on 2.1.2.0-0.2.rc2

**Comment by h0nIg** [comment]
bug was introduced by 73a3a5839b7698a6b2d6ea459b0f65234729627b (stable-2.1 branch) / 02e52c319c921222fb558d5634ebddef17e0c0c0 (devel branch) and was fixed on devel with 06d4f4ad0e8f62c6b02b01833b94a0e44be008d0. but this commit is missing on stable-2.1. please fix this on 2.1 :/

@jimi-c @alikins

**Comment by jimi-c** [comment]
That patch was extremely invasive, and will not be back-ported to stable-2.1. I have looked into other ways of fixing this without success, as it is closely tied to some of the other performance changes as well.

**Comment by jimi-c** [comment]
Due to the fact that the fix for this was closely tied to the performance improvements in 2.2, as noted above this fix will not be backported to 2.1.x. Anyone who is hitting this bug should upgrade to 2.2. If you're running into problems related to that upgrade which are holding you back, please note them here and I will make sure they get priority.

Thanks!
